### DISCUSSION TITLE
Trouble generating a 15-20KHz PWM signal on PA3 of an ATTiny404

### DISCUSSION THREAD

**Discussion by NickLD** [SpenceKonde/megaTinyCore#741] [Q&A] [answered]
Hi all.
I've been trying for the past few hours to generate a variable duty PWM signal in the 15-20KHz range on the PA3 pin of an ATTiny404 and have been quite unsuccessful. The output of this pin goes into a a gate driver IC to eventual vary the output of a DC motor.
I have tested that I'm able to get the standard ~1KHz output working with the conventional `analogWrite()` method, however this is obviously sub-optimal for motor driving.
This is my Setup:

- I am running the MCU at 16Mhz.
- I'm trying to retain `millis()` capability, so i've selected `RTC (no micros)` option in the Arduino IDE [Which is another problem in itself, as I'm unable to make this option work in my primary PlatformIO environment with specifying `DMILLIS_USE_TIMERRTC` in the `build_flags`, but that is out of the scope of this issue]

I've been trying to decipher the examples in [Taking over TCA0](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/TakingOverTCA0.md) in the READEME. It seems to me that `WO3` matches the `TCA0` timer for `PA3` according to the datasheet.
![image](https://user-images.githubusercontent.com/7613292/179148806-c16de3fe-aceb-4b48-ba92-7efbf9b0f034.png)

I've tried example 3 & the example code under "PORTMUX and accessing alternate pins" specifically for `WO3`, however with all these examples I couldn't seem to get any signal out.

I'm probably missing something obvious here. My embedded development experience doesn't go too much deeper than Arduino APIs and framework wrappers.

If anyone could give me so pointers as to what I may be doing wrong here, I would greatly appreciate it. 
Thanks!

**Comment by mechatroniks-git** [comment]
Look at this code, doing 20khz pwm. 

https://github.com/tetrabiodistributed/papra-pcb-firmware/blob/main/papracode/papracode.ino

**Reply by NickLD** [reply]
Hi, I tried to reuse your code pertaining to the PWM generation, but I still was unable to get it to work. Is there just an issue with doing this on PA3 for ATTiny404/0-series?

**Reply by NickLD** [reply]
By using this code I derived from your example:
```
void setup() {
  // put your setup code here, to run once:
    //See http://ww1.microchip.com/downloads/en/Appnotes/TB3217-Getting-Started-with-TCA-DS90003217.pdf
  TCA0.SINGLE.CTRLB = ( TCA_SINGLE_CMP2EN_bm  | TCA_SINGLE_WGMODE_SINGLESLOPE_gc ); //Single PWM on WO0 singleslope
  TCA0.SINGLE.PER   = 0xFF; // Count down from 255 on WO0/WO1/WO2
  TCA0.SINGLE.CTRLA = ( TCA_SINGLE_CLKSEL_DIV4_gc | TCA_SINGLE_ENABLE_bm ); //enable the timer with prescaler of 4
  pinMode(PIN_PB0, OUTPUT);  //PB0 - TCA0 WO0, pin9 on 14-pin parts
}

void loop() {
  // put your main code here, to run repeatedly:
  analogWrite(PIN_PB0, 128);
}
```

I was able to generate the the high-frequency PWM on `PB0`, However again I need this output on `PA3`

**Comment by grandaspanna** [comment]
According to the datasheet, in section 20.2.2, WO3 is only available in split mode. Not sure how this will impact any other uses you may have for TCA. Section 20.3.3.6 then describes split mode configuration, where *I think*, to get output on PA3, which is WO3, you need to set HCMP0 to your compare value, as well as set clock speed/scaler as appropriate. If you ran the main clock at 16MHz and used a 1024 divider, you'd end up close to 15kHz. Note that the split mode will only give you 8-bits of resolution on your PWM output.

The following code fragments are the kind of thing you may be wanting to do:
```
TCA0.SPLIT.CTRLD = 1 << TCA_SPLIT_SPLITM_bp;
TCA0.SPLIT.HCMP0 = 0x80; /* Compare value of channel 0: 0x80 */
TCA0.SPLIT.CTRLA = TCA_SPLIT_CLKSEL_DIV1024_gc /* System Clock / 1024 */
	                   | 1 << TCA_SPLIT_ENABLE_bp; /* Module Enable: enabled */
```

Of course, if you're using TCA for other things and need 16-bit single mode, then this won't help.

Also, don't forget to tell the core you're taking over TCA.

**Reply by mechatroniks-git** [reply]
I Initially had to split, but moved the pin to avoid. 

See split code here at 20khz

https://github.com/tetrabiodistributed/PAPRA-PCB/blob/ba98b61a9e022bfbdcdcd5039c62c5fe131a773c/firmware/papracode/papracode.ino

**Comment by SpenceKonde** [comment, marked as answer]
Most of what you want is already set up. There's just a small tweak you need to do (unnecessary steps commented out

```
//takeOverTCA0(); // this is needed if you change the portmux, or run in SINGLE mode, but not split mode on the default pins, 
//TCA0.SPLIT.CTRLD = 1; //Needed only if you had to take over the timer
//TCA0.SPLIT.HPER = 254; // Not 255 if using analog write! You want 256 levels of pwm, right? And 100% on and 0% on should be one of those. calling analogWrite(pin,0) or analogWrite(pin,255) is caught and calls digitalWrite on the pin, If you have the period at 255, there are 256 timer ticks per cycle, but 257 possible duty cycles including 100% and 0%.. I forget which end the discontinuity appeared on, but it would either go from 0/256ths to 2/256ths or from 254/256ths to 256/256ths.
// In fact, if you just want to use analogWrite() but have higher frequency, all you need to do with megaTinyCore, starting from the default configuration is just
TCA0.SPLIT.CTRLA=TCA_SPLIT_CLKSEL_DIV4_gc | TCA_SPLIT_ENABLE_bm;
// Now analogWrite() on any of the PWM pins of the 404 will give PWM at a speed of 15600ish Hz if system clock is 16 MHz and 19500 Hz or so if the system clock is 20 MHz..
```

Y'all are making this way more complicated than it had to be ;-)

**Reply by NickLD** [reply]
Doh! I came up with the same one-liner just a few hours ago and was planning on marking it the solution. Thanks for your help and your core! My biggest concern was if it was bad practice to not call `takeOverTCA0()` even though we were modifying its parameters. I tried with and without it, and with taking it over, it would fail. But, If you says its fine then that's all that matters to me :)
